### DISCUSSION TITLE
Problem with kong's keep alive (api gateway)

### DISCUSSION THREAD

**Discussion by ali-ghazi78** [Kong/kong#8158] [Help] [answered]
I have several k8s worker nodes that each of them run a kong instance. I also have 3 upstream servers. One month ago the number of my threads on the upstream servers increased dramatically, therefore, I was wondering about setting up keep alive in kong. I set the following variable for them:
```
KONG_UPSTREAM_KEEPALIVE_POOL_SIZE = 24
KONG_UPSTREAM_KEEPALIVE_MAX_REQUESTS = 100
KONG_UPSTREAM_KEEPALIVE_IDLE_TIMEOUT = 60
```
I set connection pool to 24 but I still have 200 thread opens on each of my upstream servers. why is that? is not it supposed to open one thread for each connection? or does it open a thread for each of requests? what is the maximum number of request that I can handle from the mentioned variables? is there any other configuration for decreasing the number of open threads on my upstream servers?

**Comment by flrgh** [comment, marked as answer]
Hello @ali-ghazi78 :wave: 

Keepalive pools are divided into buckets based on $ip+$port (or $sni+$ip+$port if using TLS), so each of your upstream servers will at very least count as one unique pool (even if they are all behind one Kong service or Kong upstream object).

So if all 5 Kong instances are configured to talk to all 24 upstream servers, you should expect 120 (5 * 24) connections per upstream server (more if there are multiple listening ports or multiple SNIs for TLS). As for why there are 200 (assuming single port and no TLS or single SNI), that's hard to comment on without knowing more about the environment. It could be that the upstream server isn't properly shutting down connections on its side after Kong closes them (due to idle timeout or max requests).

Hope this helps! Let us know if you have more questions.

**Reply by ali-ghazi78** [reply]
@flrgh Thanks for your detailed answer. However, I did not understand one part of your answer. You said that I am supposed to have 120 connections per server?? I think that you meant that I am supposed to have 120 connections on all the server as a whole ( not only on one server)

**Reply by flrgh** [reply]
The limit is 24, and it applies _per_ $IP:$PORT, _per_ Kong instance, so each of your upstream servers will have as many as 24 connections from _each_ Kong node. Since you have 5 Kong nodes this will be 24 * 5.

Here is a diagram to help illustrate. In the example there are 2 Kong instances and an upstream with 3 target servers. The pool size is set to 2, which means that each Kong instance will open 2 keepalive connections to each upstream target. Observe that each upstream server has 4 connections in total (pool size * number of kong instances):

![Screenshot from 2021-12-13 09-53-21](https://user-images.githubusercontent.com/3277009/145863333-c307f9bd-4571-49cb-9bd1-58c8277d8c8a.png)

**Reply by ali-ghazi78** [reply]
Thank you @flrgh. may I know the reference of the image? or you made it yourself? I am really thankful that you paid attention to my question. Could you please give me some more information on KONG_UPSTREAM_KEEPALIVE_MAX_REQUESTS? does it open a thread for each of requests? or they are not handled concurrently? I want to know how I can estimate the number of open threads on per upstream server.

**Reply by flrgh** [reply]
It's just something I threw together real quick on a [diagramming web app](https://app.diagrams.net/) to illustrate things a bit :)

Kong's [upstream_keepalive_max_requests setting](https://docs.konghq.com/gateway/2.6.x/reference/configuration/#upstream_keepalive_max_requests) is used to populate the NGINX upstream [keepalive_requests setting](https://nginx.org/en/docs/http/ngx_http_upstream_module.html#keepalive_requests) when Kong generates its NGINX config files (this happens when you run `kong prepare`).


Once a TCP keepalive connection to an upstream server/target has been established, the upstream_keepalive_max_requests setting controls how many requests in total that Kong/NGINX will send over that connection (so long as it remains alive/healthy). Once the maximum is reached, NGINX gracefully closes the connection, and a new connection may be established with the next request.

NGINX doesn't create a thread per connection but uses an [event loop model](https://nginx.org/en/docs/dev/development_guide.html#event_loop), so there are no new threads created when connecting to an upstream or sending a request to it. Thus, for NGINX the overhead of maintaining lots and lots of open connections to an upstream server is very minimal. On the other hand, if your upstream application creates a new thread for each established connection, then I suspect it will be beneficial for you to lower the value of `upstream_keepalive_pool_size` to reduce the total number of connections. In general you can estimate the number of open connections to your upstreams using the formula I presented of `{{number_of_kong_instances}} * {{upstream_keepalive_pool_size}}`. My advice would be to set the pool size to 1 and gradually bring it up until things are properly tuned/dialed in.

**Reply by ali-ghazi78** [reply]
Dear @flrgh, I use the following command on my upstream servers to get the number of open threads
``` lsof -f | grep PORT_NUMBER | wc -l ```
Do you have any Idea how I can investigate the kong's logs on my workers to find the number of threads that it opens on the upstream server?  
Am I using the right command to calculate the number of threads?

**Reply by flrgh** [reply]
NGINX does not expose any relatively easy means of determining why it made #N connections, and as far as I can tell, setting the keepalive pool size does not necessarily guarantee that NGINX will never create more connections. Turning your error_log level (`KONG_LOG_LEVEL`) up to `debug`, may help to give you more info or insight. You could also add more instrumentation to your upstream app to log when it receives a connection, how many requests are sent over that connection, and when a connection is closed.

The NGINX documentation may also be of help here:
https://nginx.org/en/docs/http/ngx_http_upstream_module.html
https://nginx.org/en/docs/http/ngx_http_proxy_module.html

> Am I using the right command to calculate the number of threads?

It's possible that Kong/NGINX is shutting down connections from its side, but the upstream is not properly closing them. `lsof` will show you the connection state in the rightmost column. I typically use `lsof -nP -i TCP -s TCP:^LISTEN` to show all TCP connections that are not in the "LISTEN" state.

**Comment by Tieske** [comment]
Note: didn't read the whole thread, only the OP.

That's just not how keealive pools work. They do not limit the maximum number of connections. Only the pool size.
The pool is used whenever a connection is complete, then th econnection is returned to the pool. If the pool already has 24 connections in it, then the connection isn't added, but closed instead.

When a new connection is created to the upstream, first the pool is tried, if a connection is still available, we reuse it. If not, we create one from scratch. From this you can deduce that the total number of connections is not limited by the keepalive pool size.
